### PR TITLE
Vmap decompile

### PR DESCRIPTION
**Pull request by kristiker**
image

image

### CODE REVIEW

File: GUI/Types/Viewers/Resource.cs [hunk 1]
```diff
@@ -427,6 +427,28 @@ static void VcsShaderResourceBridge(TabControl resTabs, SboxShader sboxShader)
                             VcsShaderResourceBridge(resTabs, shaderFileContainer);
                             IViewer.AddContentTab<Func<string>>(resTabs, extract.GetVfxFileName(), extract.ToVFX, true);
                             break;
+
+                        case ResourceType.Map:
+                            var mapExtract = new MapExtract(resource, vrfGuiContext.FileLoader);
+                            IViewer.AddContentTab(resTabs, $"{resource.FileName}.vmap", mapExtract.ToValveMap(), true);
+                            break;
+
+                        case ResourceType.Model:
```
Comment (kristiker) [inline GUI/Types/Viewers/Resource.cs:436]: Remove these

File: ValveResourceFormat/IO/FileExtract.cs [hunk 2]
```diff
@@ -98,7 +128,8 @@ public static ContentFile Extract(Resource resource, IFileLoader fileLoader)
             switch (resource.ResourceType)
```
Comment (kristiker) [inline ValveResourceFormat/IO/FileExtract.cs:128]: Also ModelExtract here

File: ValveResourceFormat/IO/MapExtract.cs [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+using ValveResourceFormat.Utils;
+using ValveResourceFormat.ResourceTypes;
+using System.Numerics;
+using System;
+using System.Text;
+using ValveResourceFormat.Serialization;
+using ValveResourceFormat.Blocks;
+using System.Globalization;
+using System.Linq;
+using System.Diagnostics;
```
Comment (kristiker) [inline ValveResourceFormat/IO/MapExtract.cs:13]: Order the usings alphabetically

File: ValveResourceFormat/IO/MapExtract.cs [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+        var vmapResource = FileLoader.LoadFile(vmapPath);
+        if (vmapResource is null)
+        {
+            throw new FileNotFoundException($"Failed to find vmap_c resource at {vmapPath}");
+        }
+
+        InitMapExtract(vmapResource);
+    }
+
+    private static bool PathIsSubPath(string equalOrSubPath, string path)
```
Comment (kristiker) [inline ValveResourceFormat/IO/MapExtract.cs:105]: Outdated argument var name

File: ValveResourceFormat/IO/MapExtract.cs [hunk 5]
```diff
@@ -0,0 +1,707 @@
...
+    {
+        var texture = ToolTextureMultiTags.FirstOrDefault(x => x.Value.SetEquals(combo.InteractAsStrings)).Key;
+        var tag = combo.InteractAsStrings.FirstOrDefault();
+        texture ??= tag switch
+        {
+            "playerclip" or "npcclip" => tag,
+            "sky" => "skybox",
+            "csgo_grenadeclip" => "grenadeclip",
+            _ => "nodraw",
+        };
```
Comment (kristiker) [inline ValveResourceFormat/IO/MapExtract.cs:210]: Add a note for these in MANUAL_DEFINITIONS.md

File: ValveResourceFormat/IO/ModelExtract.cs [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+        if (physAggregateData == null)
+        {
+            return;
+        }
+
+        PhysicsSurfaceNames = physAggregateData.SurfacePropertyHashes.Select(hash =>
+        {
+            StringToken.InvertedTable.TryGetValue(hash, out var name);
+            return name ?? hash.ToString(CultureInfo.InvariantCulture);
+        }).ToArray();
```
Comment (kristiker) [inline ValveResourceFormat/IO/ModelExtract.cs:336]: Populate StringToken with names from surfaceproperties.vsurf for better hit rate. Alternatively the world_physics.vmdl RED2 has the used properties listed as well.

File: ValveResourceFormat/IO/MapExtract.cs [hunk 7]
```diff
@@ -0,0 +1,707 @@
...
+        foreach (var compiledEntity in entityLump.GetEntities())
+        {
+            FixUpEntityKeyValues(compiledEntity);
+
+            var className = compiledEntity.GetProperty<string>(CommonHashes.ClassName);
+
+            if (className == "worldspawn")
+            {
+                AddProperties(compiledEntity, MapDocument.World);
+                MapDocument.World.EntityProperties["description"] = $"Decompiled with VRF 0.3.2 - https://vrf.steamdb.info/";
```
Comment (kristiker) [inline ValveResourceFormat/IO/MapExtract.cs:528]: How do I get the library version number?

Comment (kristiker) [reply ValveResourceFormat/IO/MapExtract.cs:528]: typeof(class).Assembly.GetName().Version.ToString()

**Comment by xPaw** [comment]
Merged.

**Pull request closed without merging**